p2p/peer_behaviour.go

```diff
@@ -0,0 +1,89 @@
+package p2p
+
+type ErrorBehaviourPeer int
+
+const (
+	ErrorBehaviourUnknown = iota
+	ErrorBehaviourBadMessage
+	ErrorBehaviourMessageOutofOrder
+)
+
+type GoodBehaviourPeer int
+
+const (
+	GoodBehaviourVote = iota
+	GoodBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peer Peer, reason GoodBehaviourPeer)
+	Errored(peer Peer, reason ErrorBehaviourPeer)
+}
+
+type switchedPeerBehaviour struct {
+	sw *Switch
+}
+
+func (spb *switchedPeerBehaviour) Errored(peer Peer, reason ErrorBehaviourPeer) {
+	spb.sw.StopPeerForError(peer, reason)
+}
+
+func (spb *switchedPeerBehaviour) Behaved(peer Peer, reason GoodBehaviourPeer) {
+	spb.sw.MarkPeerAsGood(peer)
+}
+
+func NewswitchedPeerBehaviour(sw *Switch) PeerBehaviour {
+	return &switchedPeerBehaviour{
+		sw: sw,
+	}
+}
+
+type ErrorBehaviours map[Peer][]ErrorBehaviourPeer
+type GoodBehaviours map[Peer][]GoodBehaviourPeer
+
+type IStorePeerBehaviour interface {
+	PeerBehaviour
+	GetErrored() ErrorBehaviours
+	GetBehaved() GoodBehaviours
+}
+
+// StorePeerBehaviour serves a mock concrete implementation of the
+// PeerBehaviour interface used in reactor tests to ensure reactors
+// produce the correct signals in manufactured scenarios.
+type storePeerBehaviour struct {
```

Comment:
As with our design method on this will be called concurrently and we therefore need to evaluate if in it's current is thread-safe and consistent.

Reply:
Indeed. The target benefactor of this interface is the `reactor.Receive` method which is called within multiple go routines. I've placed a coarse lock around both write methods which should be sufficient for correctness of the `Behaved` and `Errored` methods. However, the `GetBehaved` and `GetErrored` methods return the underlying map, which exposes the struct to a separate write path in which the receive takes the map and mutates it willy-nilly. Given that this is specifically for testing, is it sufficient to specify these limitations in a doc string or should we deep copy the structure before returning it?

Reply:
As discussed offline, changing the API to only return errors for a specific peer id will be cleaner and less error prone to implement.